### PR TITLE
GraphQL: Fix consistency and paging options

### PR DESCRIPTION
**Pull request by mpenick**

### CODE REVIEW

File: graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java [hunk 1]
```diff
@@ -25,8 +30,30 @@ public final ResultT get(DataFetchingEnvironment environment) throws Exception {
...
+      if (pageSize != null) {
+        parameters.pageSize(Integer.valueOf((String)pageSize));
+      }
+      Object pageState = options.get("pageState");
+      if (pageSize != null) {
+        parameters.pagingState(PagingState.fromString((String)pageState).getRawPagingState());
+      }
+    }
+
+    DataStore dataStore = DataStore.create(persistence, storedCredentials.getRoleName(), parameters.build());
```
Comment (olim7t) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java:56]: Nit: you could avoid allocating the builder if there are no options.
```java
    Parameters parameters;
    if (environment.containsArgument("options")) {
      ImmutableParameters.Builder builder = Parameters.builder();
      ...
      parameters = builder.build();
    } else {
      parameters = Parameters.defaults();
    }
```

**Review by olim7t — COMMENTED** [review]
👍 
You'll need to update the mocking in `GraphQlTestBase`.
I think we should also add an `expectedParameters` argument to `GraphQlTestBase.assertSuccess`, that will allow you to cover those changes in unit tests.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java (hunk 1 above)
Comment (mpenick) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java:56]: Thanks, good suggestion.

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
           

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java  5
         

Complexity decreasing per file
==============================
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/QueryFetcher.java  -1
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/DeleteMutationFetcher.java  -4
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/UpdateMutationFetcher.java  -6
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/InsertMutationFetcher.java  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6340978&bid=20597213)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
           

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java  5
         

Complexity decreasing per file
==============================
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/UpdateMutationFetcher.java  -6
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/InsertMutationFetcher.java  -4
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/DeleteMutationFetcher.java  -4
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/QueryFetcher.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6340978&bid=20597213)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
           

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java  5
         

Complexity decreasing per file
==============================
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/UpdateMutationFetcher.java  -6
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/DeleteMutationFetcher.java  -4
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/QueryFetcher.java  -1
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/InsertMutationFetcher.java  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6340978&bid=20597213)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
           

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java  5
         

Complexity decreasing per file
==============================
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/DeleteMutationFetcher.java  -4
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/QueryFetcher.java  -1
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/InsertMutationFetcher.java  -4
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/UpdateMutationFetcher.java  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6340978&bid=20597213)

File: graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java [hunk 2]
```diff
@@ -25,8 +31,39 @@ public final ResultT get(DataFetchingEnvironment environment) throws Exception {
...
+        builder.consistencyLevel(ConsistencyLevel.valueOf((String) consistency));
+      }
+
+      Object serialConsistency = options.get("serialConsistency");
+      if (serialConsistency != null) {
+        builder.serialConsistencyLevel(ConsistencyLevel.valueOf((String) serialConsistency));
+      }
+
+      Object pageSize = options.get("pageSize");
+      if (pageSize != null) {
```
Comment (tomekl007) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java:51]: shouldn't we also that is instanceof Integer?
` if (pageSize != null && pageSize instanceof Integer)`
(similar for other checks)?

Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java:51]: The GraphQL parser will make sure the correct type is provided, so there's no need for additional validation.

**Review by jorgebay — APPROVED** [review]
👍 

Not necessary related to this patch:  I don't usually suggest moving string literals to constants, but I think GraphQL module has too many string literals :) 

We could extract constants in the future.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/QueryFetcher.java [hunk 3]
```diff
@@ -52,11 +54,15 @@ public QueryFetcher(
...
+
     return ImmutableMap.of(
         "values",
-        resultSet.rows().stream()
+        resultSet.currentPageRows().stream()
             .map(row -> DataTypeMapping.toGraphQLValue(nameMapping, table, row))
-            .collect(Collectors.toList()));
+            .collect(Collectors.toList()),
+        "pageState",
+        pageState == null ? "" : Base64.getEncoder().encodeToString(pageState.array()));
```
[the remark this answers is not in the document]
Comment (mpenick) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/QueryFetcher.java:65]: I used `null` first, but it causes a NPE.

[the remark this answers is not in the document]
Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/QueryFetcher.java:65]: oh, gtk!

[the remark this answers is not in the document]
Comment (pcmanus) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/QueryFetcher.java:65]: Yes, `ImmutableMap` does not like `null`, but I think wouldn't it be cleaner to just not add an entry in the map if there is no pageState. Using an empty string means consumer now have to check for both `null` _and_ the empty string when getting from that map.

To be fair, that resulting map seems to be used by `graphql-java` and I'm not sure how exactly, so maybe there is a good reason we _want_ a "pageState" entry in that map. But if so, that's subtle, and I strongly suggest a comment.

File: graphqlapi/src/test/java/io/stargate/graphql/schema/fetchers/dml/QueryFetcherTest.java [hunk 4]
```diff
@@ -50,4 +57,32 @@ public void errorTest(String graphQlMutation, String expectedError) {
...
   }
+
+  @ParameterizedTest
+  @MethodSource("operationsWithOptions")
+  @DisplayName("Should execute GraphQL with options passed correctly")
+  public void consistencyAndPagingOptionsTest(
+      String graphQlOperation, Parameters expectedParameters) {
+    ExecutionResult result = executeGraphQl(graphQlOperation);
+    assertThat(result.getErrors()).isEmpty();
+    assertThat(parametersCaptor.getValue()).isEqualTo(expectedParameters);
```
Comment (olim7t) [inline graphqlapi/src/test/java/io/stargate/graphql/schema/fetchers/dml/QueryFetcherTest.java:68]: Good idea to make this a separate test.

We might want to extract this logic to a parent class like `assertSuccess`, but we can do it later when needed.

**Review by olim7t — APPROVED** [review]

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
           

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java  5
         

Complexity decreasing per file
==============================
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/QueryFetcher.java  -1
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/UpdateMutationFetcher.java  -6
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/DeleteMutationFetcher.java  -4
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/InsertMutationFetcher.java  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6340978&bid=20597213)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
           

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java  5
         

Complexity decreasing per file
==============================
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/InsertMutationFetcher.java  -4
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/QueryFetcher.java  -1
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/UpdateMutationFetcher.java  -6
+ graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/DeleteMutationFetcher.java  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6340978&bid=20597213)

**Pull request merged**